### PR TITLE
Update theme resource values when system theme changes

### LINKED ISSUE #3300: System theme changes are not applied
## Current behavior

When OS theme changes at runtime, the `Application.OnRequestedThemeChanged` method executes, but the theme does not apply.

## Expected behavior

Theme should change.

## How to reproduce it (as minimally and precisely as possible)

- Replace SamplesApp MainPage with:

```
  <Grid Background="{ThemeResource ApplicationPageBackgroundThemeBrush}">
	<TextBlock Text="Hello, world!" Margin="20" FontSize="30" />
  </Grid>
```
- Add a debug message to the start of the `OnRequestedThemeChanged` method
- Remove `RequestedTheme` attribute from `Application.xaml`
- Run the app and change theme at runtime. Observe the theme change is not applied

## Environment

Nuget Package: dev

Package Version(s): dev

Affected platform(s):

- [x] iOS
- [x] Android
- [x] WebAssembly
- [ ] WebAssembly renderers for Xamarin.Forms
- [x] macOS
- [ ] Windows
- [ ] Build tasks
- [ ] Solution Templates

Visual Studio:

- [ ] 2017 (version: )
- [ ] 2019 (version: )
- [ ] for Mac (version: )

Relevant plugins:

- [ ] Resharper (version: )

### PR DESCRIPTION
**Pull request by davidjohnoliver**
GitHub Issue (If applicable): fixes #3300 

## PR Type

- Feature

What kind of change does this PR introduce?

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/Setter.cs [hunk 1]
```diff
@@ -74,7 +84,14 @@ internal override void ApplyTo(DependencyObject o)
 			if (Property != null)
 			{
 				object value = _valueProvider != null ? _valueProvider() : _value;
-				o.SetValue(Property, BindingPropertyHelper.Convert(() => Property.Type, value));
+				if (!ThemeResourceName.IsNullOrEmpty())
```
Comment (dr1rrb) [inline src/Uno.UI/UI/Xaml/Setter.cs:87]: You should put the positive condition first, it's more readable :)

File: src/Uno.UI/UI/Xaml/Style/Style.cs [hunk 2]
```diff
@@ -67,6 +67,9 @@ internal void ApplyTo(DependencyObject o, DependencyPropertyValuePrecedences pre
 					{
 						pair.Value(o);
 					}
+
+					// Check tree for resource binding values, since some Setters may have set ThemeResource-backed values
+					(o as IDependencyObjectStoreProvider).Store.UpdateResourceBindings(isThemeChangedUpdate: false);
```
Comment (dr1rrb) [inline src/Uno.UI/UI/Xaml/Style/Style.cs:72]: I like the readability of the safe cast, but a null ref in logs is much less helpful than an invalid cast exception which dumps the source and target type ...

Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/Style/Style.cs:72]: In this case `DependencyObject` is guaranteed to be a `IDependencyObjectStoreProvider` by the backing generator, the safe cast is just to access the `Store` property (which is explicitly implemented)

**Pull request merged**